File: backend/services/route_optimizer.py
"""
동선 최적화 모듈 - TSP with 2-opt + 지역 클러스터링

최적화 전략:
1. 지역별 그룹핑 (제주시, 서귀포, 동부, 서부, 중산간)
2. 최적 지역 순서 결정 (역주행 방지)
3. 지역 내 Nearest Neighbor + 2-opt
4. 효율성 점수 계산
"""

from typing import Any
from .jeju_regions import (
    haversine_distance,
    classify_place_by_region,
    group_places_by_region,
    get_optimal_region_order,
    calculate_region_order_score,
)
from models.schemas import SchedulePlace, DaySchedule, RouteEfficiency


def estimate_travel_time(dist_km: float, has_rentcar: bool) -> int:
    """이동 시간 추정 (분)"""
    speed_kmh = 40 if has_rentcar else 25
    base_minutes = (dist_km / speed_kmh) * 60
    # 대중교통은 대기시간 10분 추가
    return round(base_minutes if has_rentcar else base_minutes + 10)


def reorder_places_nearest_neighbor(places: list[dict]) -> list[dict]:
    """Nearest-neighbor로 장소 재배열 (첫 번째 장소 고정)"""
    if len(places) <= 2:
        return places

    result = [places[0]]
    remaining = places[1:]

    while remaining:
        last = result[-1]
        nearest_idx = 0
        nearest_dist = float("inf")

        for i, place in enumerate(remaining):
            dist = haversine_distance(
                last["latitude"],
                last["longitude"],
                place["latitude"],
                place["longitude"],
            )
            if dist < nearest_dist:
                nearest_dist = dist
                nearest_idx = i

        result.append(remaining.pop(nearest_idx))

    return result


def two_opt_optimize(places: list[dict]) -> list[dict]:
    """2-opt 알고리즘으로 경로 개선"""
    if len(places) <= 3:
        return places

    route = places.copy()
    improved = True
    iterations = 0
    max_iterations = 100

    while improved and iterations < max_iterations:
        improved = False
        iterations += 1

        for i in range(1, len(route) - 2):
            for j in range(i + 1, len(route) - 1):
                # 현재 거리
                current_dist = haversine_distance(
                    route[i - 1]["latitude"],
                    route[i - 1]["longitude"],
                    route[i]["latitude"],
                    route[i]["longitude"],
                ) + haversine_distance(
                    route[j]["latitude"],
                    route[j]["longitude"],
                    route[j + 1]["latitude"],
                    route[j + 1]["longitude"],
                )

                # 뒤집었을 때 거리
                new_dist = haversine_distance(
                    route[i - 1]["latitude"],
                    route[i - 1]["longitude"],
                    route[j]["latitude"],
                    route[j]["longitude"],
                ) + haversine_distance(
                    route[i]["latitude"],
                    route[i]["longitude"],
                    route[j + 1]["latitude"],
                    route[j + 1]["longitude"],
                )

                # 개선되면 구간 뒤집기
                if new_dist < current_dist - 0.1:
                    route[i : j + 1] = reversed(route[i : j + 1])
                    improved = True

    return route


def calculate_total_distance(places: list[dict]) -> float:
    """총 이동거리 계산 (km)"""
    if len(places) < 2:
        return 0

    total = 0
    for i in range(len(places) - 1):
        total += haversine_distance(
            places[i]["latitude"],
            places[i]["longitude"],
            places[i + 1]["latitude"],
            places[i + 1]["longitude"],
        )
    return total


def detect_backtracking(places: list[dict]) -> dict:
    """역주행 감지"""
    details = []

    if len(places) < 3:
        return {"count": 0, "details": details}

    regions = [classify_place_by_region(p["latitude"], p["longitude"]) for p in places]
    visited_regions: set[str] = set()
    last_region = regions[0]
    visited_regions.add(last_region)

    for i in range(1, len(regions)):
        current_region = regions[i]
        if current_region != last_region:
            if current_region in visited_regions:
                details.append({
                    "fromIdx": i - 1,
                    "toIdx": i,
                    "region": current_region,
                })
            visited_regions.add(current_region)
            last_region = current_region

    return {"count": len(details), "details": details}


def calculate_efficiency(places: list[dict], has_rentcar: bool) -> RouteEfficiency:
    """동선 효율성 점수 계산"""
    if not places:
        return RouteEfficiency(
            totalDistance=0,
            totalTravelTime=0,
            regionScore=100,
            backtrackCount=0,
            efficiencyScore=100,
        )

    total_distance = calculate_total_distance(places)
    total_travel_time = sum(p.get("travelTime", 0) for p in places)

    regions = [classify_place_by_region(p["latitude"], p["longitude"]) for p in places]
    region_score = calculate_region_order_score(regions)

    backtrack_info = detect_backtracking(places)
    backtrack_count = backtrack_info["count"]
    backtrack_penalty = min(backtrack_count * 10, 30)

    # 거리 효율성
    distance_efficiency = 100
    if len(places) >= 2:
        direct_distance = haversine_distance(
            places[0]["latitude"],
            places[0]["longitude"],
            places[-1]["latitude"],
            places[-1]["longitude"],
        )
        if direct_distance > 0:
            ratio = total_distance / direct_distance
            distance_efficiency = max(0, 100 - max(0, ratio - 2) * 20)

    efficiency_score = round(
        region_score * 0.5 + (100 - backtrack_penalty) * 0.3 + distance_efficiency * 0.2
    )

    return RouteEfficiency(
        totalDistance=round(total_distance, 1),
        totalTravelTime=total_travel_time,
        regionScore=region_score,
        backtrackCount=backtrack_count,
        efficiencyScore=max(0, min(100, efficiency_score)),
    )


def optimize_with_regions(places: list[dict]) -> list[dict]:
    """지역 기반 최적화"""
    if len(places) <= 2:
        return places

    grouped = group_places_by_region(places)
    regions = list(grouped.keys())

    if len(regions) <= 1:
        nn_result = reorder_places_nearest_neighbor(places)
        return two_opt_optimize(nn_result)

    start_region = classify_place_by_region(places[0]["latitude"], places[0]["longitude"])
    optimal_order = get_optimal_region_order(start_region, regions)

    result: list[dict] = []

    for region in optimal_order:
        if region in grouped and grouped[region]:
            region_places = grouped[region]

            if result:
                last_place = result[-1]
                nearest_idx = 0
                nearest_dist = float("inf")

                for i, place in enumerate(region_places):
                    dist = haversine_distance(
                        last_place["latitude"],
                        last_place["longitude"],
                        place["latitude"],
                        place["longitude"],
                    )
                    if dist < nearest_dist:
                        nearest_dist = dist
                        nearest_idx = i

                nearest = region_places[nearest_idx]
                region_places = [nearest] + [
                    p for i, p in enumerate(region_places) if i != nearest_idx
                ]

            nn_result = reorder_places_nearest_neighbor(region_places)
            optimized = two_opt_optimize(nn_result)
            result.extend(optimized)

    # 포함되지 않은 지역 추가
    for region in regions:
        if region not in optimal_order and region in grouped:
            result.extend(grouped[region])

    return result


def parse_time(time_str: str) -> int:
    """시간 문자열 파싱 → 분"""
    parts = time_str.split(":")
    return int(parts[0]) * 60 + int(parts[1])


def format_time(minutes: int) -> str:
    """분 → 시간 문자열"""
    h = minutes // 60
    m = minutes % 60
    return f"{h:02d}:{m:02d}"


def recalculate_times(places: list[dict], has_rentcar: bool) -> list[dict]:
    """재배열 후 시간 재계산 + travelTime 설정"""
    if not places:
        return places

    result = [p.copy() for p in places]
    current_end = parse_time(result[0]["time"]) + result[0].get("duration", 60)

    for i in range(len(result) - 1):
        dist = haversine_distance(
            result[i]["latitude"],
            result[i]["longitude"],
            result[i + 1]["latitude"],
            result[i + 1]["longitude"],
        )
        travel = estimate_travel_time(dist, has_rentcar)
        result[i]["travelTime"] = travel

        next_start = current_end + travel
        result[i + 1]["time"] = format_time(next_start)
        current_end = next_start + result[i + 1].get("duration", 60)

    return result


def optimize_route(schedule: list[dict], has_rentcar: bool) -> list[dict]:
    """메인 함수: 전체 스케줄 최적화"""
    optimized_schedule = []

    for day in schedule:
        places = day.get("places", [])

        # dict로 변환 (Pydantic 모델인 경우)
        places_dicts = []
        for p in places:
            if hasattr(p, "model_dump"):
                places_dicts.append(p.model_dump())
            elif isinstance(p, dict):
                places_dicts.append(p)
            else:
                places_dicts.append(dict(p))

        # 지역 기반 최적화
        optimized_order = optimize_with_regions(places_dicts)

        # 시간 재계산
        with_times = recalculate_times(optimized_order, has_rentcar)

        optimized_schedule.append({
            "day": day.get("day"),
            "date": day.get("date"),
            "places": with_times,
        })

    return optimized_schedule


def analyze_schedule_efficiency(
    schedule: list[dict], has_rentcar: bool
) -> dict:
    """전체 스케줄의 효율성 분석"""
    days = []

    for day in schedule:
        places = day.get("places", [])
        places_dicts = []
        for p in places:
            if hasattr(p, "model_dump"):
                places_dicts.append(p.model_dump())
            elif isinstance(p, dict):
                places_dicts.append(p)
            else:
                places_dicts.append(dict(p))

        efficiency = calculate_efficiency(places_dicts, has_rentcar)
        days.append(efficiency.model_dump())

    if not days:
        return {"days": [], "overall": RouteEfficiency().model_dump()}

    overall = RouteEfficiency(
        totalDistance=sum(d["totalDistance"] for d in days),
        totalTravelTime=sum(d["totalTravelTime"] for d in days),
        regionScore=round(sum(d["regionScore"] for d in days) / len(days)),
        backtrackCount=sum(d["backtrackCount"] for d in days),
        efficiencyScore=round(sum(d["efficiencyScore"] for d in days) / len(days)),
    )

    return {"days": days, "overall": overall.model_dump()}

File: lib/routeOptimizer.ts
import { DaySchedule, SchedulePlace, RouteEfficiency } from "./types";
import {
  classifyPlaceByRegion,
  groupPlacesByRegion,
  getOptimalRegionOrder,
  calculateRegionOrderScore,
} from "./jejuRegions";

/**
 * 동선 최적화 모듈 - TSP with 2-opt + 지역 클러스터링
 *
 * 최적화 전략:
 * 1. 지역별 그룹핑 (제주시, 서귀포, 동부, 서부, 중산간)
 * 2. 최적 지역 순서 결정 (역주행 방지)
 * 3. 지역 내 Nearest Neighbor + 2-opt
 * 4. 효율성 점수 계산
 */

/** Haversine 거리 계산 (km) */
function haversineDistance(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/** 이동 시간 추정 (분) */
function estimateTravelTime(distKm: number, hasRentcar: boolean): number {
  const speedKmh = hasRentcar ? 40 : 25;
  const baseMinutes = (distKm / speedKmh) * 60;
  // 대중교통은 대기시간 10분 추가
  return Math.round(hasRentcar ? baseMinutes : baseMinutes + 10);
}

/** Nearest-neighbor로 장소 재배열 (첫 번째 장소 고정) */
function reorderPlaces(places: SchedulePlace[]): SchedulePlace[] {
  if (places.length <= 2) return places;

  const result: SchedulePlace[] = [places[0]];
  const remaining = places.slice(1);

  while (remaining.length > 0) {
    const last = result[result.length - 1];
    let nearestIdx = 0;
    let nearestDist = Infinity;

    for (let i = 0; i < remaining.length; i++) {
      const dist = haversineDistance(
        last.latitude,
        last.longitude,
        remaining[i].latitude,
        remaining[i].longitude
      );
      if (dist < nearestDist) {
        nearestDist = dist;
        nearestIdx = i;
      }
    }

    result.push(remaining[nearestIdx]);
    remaining.splice(nearestIdx, 1);
  }

  return result;
}

/**
 * 거리 행렬 생성 (TSP용)
 */
function buildDistanceMatrix(places: SchedulePlace[]): number[][] {
  const n = places.length;
  const matrix: number[][] = Array(n)
    .fill(null)
    .map(() => Array(n).fill(0));

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const dist = haversineDistance(
        places[i].latitude,
        places[i].longitude,
        places[j].latitude,
        places[j].longitude
      );
      matrix[i][j] = dist;
      matrix[j][i] = dist;
    }
  }

  return matrix;
}

/**
 * 총 이동거리 계산
 */
function calculateTotalDistance(
  places: SchedulePlace[],
  distMatrix?: number[][]
): number {
  if (places.length < 2) return 0;

  let total = 0;
  for (let i = 0; i < places.length - 1; i++) {
    if (distMatrix) {
      // 원본 인덱스를 알 수 없으므로 직접 계산
      total += haversineDistance(
        places[i].latitude,
        places[i].longitude,
        places[i + 1].latitude,
        places[i + 1].longitude
      );
    } else {
      total += haversineDistance(
        places[i].latitude,
        places[i].longitude,
        places[i + 1].latitude,
        places[i + 1].longitude
      );
    }
  }

  return total;
}

/**
 * 2-opt 알고리즘으로 경로 개선
 * Nearest Neighbor 결과를 더 최적화
 */
function twoOptOptimize(places: SchedulePlace[]): SchedulePlace[] {
  if (places.length <= 3) return places;

  let route = [...places];
  let improved = true;
  let iterations = 0;
  const maxIterations = 100; // 무한루프 방지

  while (improved && iterations < maxIterations) {
    improved = false;
    iterations++;

    // 첫 번째 장소(시작점)는 고정
    for (let i = 1; i < route.length - 2; i++) {
      for (let j = i + 1; j < route.length - 1; j++) {
        const currentDist =
          haversineDistance(
            route[i - 1].latitude,
            route[i - 1].longitude,
            route[i].latitude,
            route[i].longitude
          ) +
          haversineDistance(
            route[j].latitude,
            route[j].longitude,
            route[j + 1].latitude,
            route[j + 1].longitude
          );

        // i와 j 사이를 뒤집었을 때의 거리
        const newDist =
          haversineDistance(
            route[i - 1].latitude,
            route[i - 1].longitude,
            route[j].latitude,
            route[j].longitude
          ) +
          haversineDistance(
            route[i].latitude,
            route[i].longitude,
            route[j + 1].latitude,
            route[j + 1].longitude
          );

        // 개선되면 구간 뒤집기
        if (newDist < currentDist - 0.1) {
          // 0.1km 이상 개선시
          const reversed = route.slice(i, j + 1).reverse();
          route = [...route.slice(0, i), ...reversed, ...route.slice(j + 1)];
          improved = true;
        }
      }
    }
  }

  return route;
}

/**
 * 역주행 감지 (같은 지역을 나갔다가 다시 돌아오는 경우)
 */
export function detectBacktracking(places: SchedulePlace[]): {
  count: number;
  details: { fromIdx: number; toIdx: number; region: string }[];
} {
  const details: { fromIdx: number; toIdx: number; region: string }[] = [];

  if (places.length < 3) return { count: 0, details };

  // 각 장소의 지역 분류
  const regions = places.map((p) =>
    classifyPlaceByRegion(p.latitude, p.longitude)
  );

  // 방문한 지역 추적
  const visitedRegions = new Set<string>();
  let lastRegion = regions[0];
  visitedRegions.add(lastRegion);

  for (let i = 1; i < regions.length; i++) {
    const currentRegion = regions[i];

    if (currentRegion !== lastRegion) {
      // 지역이 바뀜
      if (visitedRegions.has(currentRegion)) {
        // 이미 방문한 지역으로 돌아감 = 역주행
        details.push({
          fromIdx: i - 1,
          toIdx: i,
          region: currentRegion,
        });
      }
      visitedRegions.add(currentRegion);
      lastRegion = currentRegion;
    }
  }

  return { count: details.length, details };
}

/**
 * 동선 효율성 점수 계산
 */
export function calculateEfficiency(
  places: SchedulePlace[],
  hasRentcar: boolean
): RouteEfficiency {
  if (places.length === 0) {
    return {
      totalDistance: 0,
      totalTravelTime: 0,
      regionScore: 100,
      backtrackCount: 0,
      efficiencyScore: 100,
    };
  }

  // 총 이동거리
  const totalDistance = calculateTotalDistance(places);

  // 총 이동시간
  const totalTravelTime = places.reduce(
    (sum, p) => sum + (p.travelTime || 0),
    0
  );

  // 지역 순서 점수
  const regions = places.map((p) =>
    classifyPlaceByRegion(p.latitude, p.longitude)
  );
  const regionScore = calculateRegionOrderScore(regions);

  // 역주행 횟수
  const { count: backtrackCount } = detectBacktracking(places);

  // 종합 효율성 (가중치 적용)
  // - 지역 점수 50%
  // - 역주행 패널티 30%
  // - 거리 효율성 20%
  const backtrackPenalty = Math.min(backtrackCount * 10, 30);

  // 거리 효율성: 직선거리 대비 실제 이동거리
  let distanceEfficiency = 100;
  if (places.length >= 2) {
    const directDistance = haversineDistance(
      places[0].latitude,
      places[0].longitude,
      places[places.length - 1].latitude,
      places[places.length - 1].longitude
    );
    if (directDistance > 0) {
      const ratio = totalDistance / directDistance;
      // ratio가 1에 가까울수록 효율적 (하지만 여러 장소 방문이므로 1.5~2가 적정)
      distanceEfficiency = Math.max(0, 100 - Math.max(0, ratio - 2) * 20);
    }
  }

  const efficiencyScore = Math.round(
    regionScore * 0.5 +
      (100 - backtrackPenalty) * 0.3 +
      distanceEfficiency * 0.2
  );

  return {
    totalDistance: Math.round(totalDistance * 10) / 10,
    totalTravelTime,
    regionScore,
    backtrackCount,
    efficiencyScore: Math.max(0, Math.min(100, efficiencyScore)),
  };
}

/**
 * 지역 기반 최적화
 * 1. 장소를 지역별로 그룹핑
 * 2. 최적 지역 순서 결정
 * 3. 지역 내 Nearest Neighbor + 2-opt
 */
function optimizeWithRegions(places: SchedulePlace[]): SchedulePlace[] {
  if (places.length <= 2) return places;

  // 지역별 그룹핑
  const grouped = groupPlacesByRegion(places);
  const regions = Object.keys(grouped);

  if (regions.length <= 1) {
    // 모든 장소가 한 지역 내 → 일반 최적화
    const nnResult = reorderPlaces(places);
    return twoOptOptimize(nnResult);
  }

  // 첫 번째 장소의 지역 확인
  const startRegion = classifyPlaceByRegion(
    places[0].latitude,
    places[0].longitude
  );

  // 최적 지역 순서 결정
  const optimalOrder = getOptimalRegionOrder(startRegion, regions);

  // 지역 순서대로 장소 배열 (각 지역 내부는 2-opt 최적화)
  const result: SchedulePlace[] = [];

  for (const region of optimalOrder) {
    if (grouped[region] && grouped[region].length > 0) {
      let regionPlaces = grouped[region];

      // 이전 장소가 있으면 가장 가까운 곳부터 시작
      if (result.length > 0) {
        const lastPlace = result[result.length - 1];
        let nearestIdx = 0;
        let nearestDist = Infinity;

        for (let i = 0; i < regionPlaces.length; i++) {
          const dist = haversineDistance(
            lastPlace.latitude,
            lastPlace.longitude,
            regionPlaces[i].latitude,
            regionPlaces[i].longitude
          );
          if (dist < nearestDist) {
            nearestDist = dist;
            nearestIdx = i;
          }
        }

        // 가장 가까운 곳을 첫 번째로
        const nearest = regionPlaces[nearestIdx];
        regionPlaces = [
          nearest,
          ...regionPlaces.filter((_, i) => i !== nearestIdx),
        ];
      }

      // 지역 내 Nearest Neighbor + 2-opt
      const nnResult = reorderPlaces(regionPlaces);
      const optimized = twoOptOptimize(nnResult);
      result.push(...optimized);
    }
  }

  // 포함되지 않은 지역 추가 (기타 지역)
  for (const region of regions) {
    if (!optimalOrder.includes(region) && grouped[region]) {
      result.push(...grouped[region]);
    }
  }

  return result;
}

/** 시간 문자열 파싱 → 분 */
function parseTime(time: string): number {
  const [h, m] = time.split(":").map(Number);
  return h * 60 + m;
}

/** 분 → 시간 문자열 */
function formatTime(minutes: number): string {
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  return `${h.toString().padStart(2, "0")}:${m.toString().padStart(2, "0")}`;
}

/** 재배열 후 시간 재계산 + travelTime 설정 */
export function recalculateTimes(
  places: SchedulePlace[],
  hasRentcar: boolean
): SchedulePlace[] {
  if (places.length === 0) return places;

  const result = [...places];
  // 첫 장소는 원래 시간 유지
  let currentEnd = parseTime(result[0].time) + result[0].duration;

  for (let i = 0; i < result.length - 1; i++) {
    const dist = haversineDistance(
      result[i].latitude,
      result[i].longitude,
      result[i + 1].latitude,
      result[i + 1].longitude
    );
    const travel = estimateTravelTime(dist, hasRentcar);
    result[i] = { ...result[i], travelTime: travel };

    const nextStart = currentEnd + travel;
    result[i + 1] = { ...result[i + 1], time: formatTime(nextStart) };
    currentEnd = nextStart + result[i + 1].duration;
  }

  return result;
}

/**
 * 메인 함수: 전체 스케줄 최적화
 * 지역 클러스터링 + Nearest Neighbor + 2-opt
 */
export function optimizeRoute(
  schedule: DaySchedule[],
  hasRentcar: boolean
): DaySchedule[] {
  return schedule.map((day) => {
    // 1. 지역 기반 최적화 (클러스터링 + 2-opt)
    const optimizedOrder = optimizeWithRegions(day.places);

    // 2. 시간 재계산
    const withTimes = recalculateTimes(optimizedOrder, hasRentcar);

    return { ...day, places: withTimes };
  });
}

/**
 * 전체 스케줄의 효율성 분석
 */
export function analyzeScheduleEfficiency(
  schedule: DaySchedule[],
  hasRentcar: boolean
): {
  days: RouteEfficiency[];
  overall: RouteEfficiency;
} {
  const days = schedule.map((day) => calculateEfficiency(day.places, hasRentcar));

  // 전체 평균/합계
  const overall: RouteEfficiency = {
    totalDistance: days.reduce((sum, d) => sum + d.totalDistance, 0),
    totalTravelTime: days.reduce((sum, d) => sum + d.totalTravelTime, 0),
    regionScore: Math.round(
      days.reduce((sum, d) => sum + d.regionScore, 0) / days.length
    ),
    backtrackCount: days.reduce((sum, d) => sum + d.backtrackCount, 0),
    efficiencyScore: Math.round(
      days.reduce((sum, d) => sum + d.efficiencyScore, 0) / days.length
    ),
  };

  return { days, overall };
}

/**
 * 최적화 전후 비교
 */
export function compareRoutes(
  before: SchedulePlace[],
  after: SchedulePlace[],
  hasRentcar: boolean
): {
  before: RouteEfficiency;
  after: RouteEfficiency;
  improvement: {
    distance: number;     // km 감소
    time: number;         // 분 감소
    score: number;        // 점수 증가
  };
} {
  const beforeEff = calculateEfficiency(before, hasRentcar);
  const afterEff = calculateEfficiency(after, hasRentcar);

  return {
    before: beforeEff,
    after: afterEff,
    improvement: {
      distance: Math.round((beforeEff.totalDistance - afterEff.totalDistance) * 10) / 10,
      time: beforeEff.totalTravelTime - afterEff.totalTravelTime,
      score: afterEff.efficiencyScore - beforeEff.efficiencyScore,
    },
  };
}
